crates/bevy_math/src/curve/mod.rs

```diff
@@ -0,0 +1,665 @@
+    /// # use bevy_math::curve::*;
+    /// let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |t| t * t + 1.0);
+    /// // Borrow `my_curve` long enough to resample a mapped version. Note that `map` takes
+    /// // ownership of its input.
+    /// let samples = my_curve.by_ref().map(|x| x * 2.0).resample_auto(100).unwrap();
+    /// // Do something else with `my_curve` since we retained ownership:
+    /// let new_curve = my_curve.reparametrize_linear(interval(-1.0, 1.0).unwrap()).unwrap();
+    /// ```
+    fn by_ref(&self) -> &Self
+    where
+        Self: Sized,
+    {
+        self
+    }
+
+    /// Flip this curve so that its tuple output is arranged the other way.
+    fn flip<U, V>(self) -> impl Curve<(V, U)>
+    where
+        Self: Sized + Curve<(U, V)>,
+    {
+        self.map(|(u, v)| (v, u))
+    }
+}
+
+impl<T, C, D> Curve<T> for D
+where
+    C: Curve<T> + ?Sized,
+    D: Deref<Target = C>,
+{
+    fn domain(&self) -> Interval {
+        <C as Curve<T>>::domain(self)
+    }
+
+    fn sample_unchecked(&self, t: f32) -> T {
+        <C as Curve<T>>::sample_unchecked(self, t)
+    }
+}
+
+/// An error indicating that an end-to-end composition couldn't be performed because of
+/// malformed inputs.
+#[derive(Debug, Error)]
+#[error("Could not compose these curves together")]
+pub enum ChainError {
+    /// The right endpoint of the first curve was infinite.
+    #[error("The first curve's domain has an infinite end")]
+    FirstEndInfinite,
+
+    /// The left endpoint of the second curve was infinite.
+    #[error("The second curve's domain has an infinite start")]
+    SecondStartInfinite,
+}
+
+/// A curve with a constant value over its domain.
+///
+/// This is a curve that holds an inner value and always produces a clone of that value when sampled.
+#[derive(Clone, Copy, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
+pub struct ConstantCurve<T> {
+    domain: Interval,
+    value: T,
+}
+
+impl<T> ConstantCurve<T>
+where
+    T: Clone,
+{
+    /// Create a constant curve, which has the given `domain` and always produces the given `value`
+    /// when sampled.
+    pub fn new(domain: Interval, value: T) -> Self {
+        Self { domain, value }
+    }
+}
+
+impl<T> Curve<T> for ConstantCurve<T>
+where
+    T: Clone,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.domain
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, _t: f32) -> T {
+        self.value.clone()
+    }
+}
+
+/// A curve defined by a function together with a fixed domain.
+///
+/// This is a curve that holds an inner function `f` which takes numbers (`f32`) as input and produces
+/// output of type `T`. The value of this curve when sampled at time `t` is just `f(t)`.
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
```

Comment:
We are introducing function reflection.

Reply:
> We are introducing function reflection.

This is true, although I don't think that necessarily changes anything here short-term, since the reflection is ultimately done through `DynamicFunction`, which doesn't itself implement the required `Fn` trait (although maybe it could do something like that in the future). 

On the other hand, it's theoretically possible for users to create their own types which implement `Fn` and can be reflected or serialized, which is why I landed on including this as "covering bases" in the first place.